src/mlpack/methods/linear_svm/linear_svm_main.cpp

```diff
@@ -0,0 +1,419 @@
+      "epochs must be non-negative");
+
+  // These are the matrices we might use.
+  arma::mat trainingSet;
+  arma::Row<size_t> labels;
+  arma::Row<size_t> rawLabels;
+  arma::mat testSet;
+  arma::Row<size_t> predictions;
+  arma::Col<size_t> mappings;
+  size_t numClasses;
+
+  // Load data matrix.
+  if (CLI::HasParam("training"))
+    trainingSet = std::move(CLI::GetParam<arma::mat>("training"));
+
+  // Check if the labels are in a separate file.
+  if (CLI::HasParam("training") && CLI::HasParam("labels"))
+  {
+    rawLabels = std::move(CLI::GetParam<arma::Row<size_t>>("labels"));
+    if (trainingSet.n_cols != rawLabels.n_cols)
+    {
+      Log::Fatal << "The labels must have the same number of points as the "
+          << "training dataset." << endl;
+    }
+  }
+  else if (CLI::HasParam("training"))
+  {
+    // Checking the size of training data if no labels are passed.
+    if (trainingSet.n_rows < 2)
+    {
+      Log::Fatal << "Can't get labels from training data since it has less "
+          << "than 2 rows." << endl;
+    }
+
+    // The initial predictors for y, Nx1.
+    rawLabels = arma::conv_to<arma::Row<size_t>>::from(
+        trainingSet.row(trainingSet.n_rows - 1));
+    trainingSet.shed_row(trainingSet.n_rows - 1);
+  }
+
+  data::NormalizeLabels(rawLabels, labels, mappings);
+
+  // Load the model, if necessary.
+  LinearSVM<>* model;
+  if (CLI::HasParam("input_model"))
+  {
+    model = CLI::GetParam<LinearSVM<>*>("input_model");
+  }
+  else
+  {
+    numClasses = CLI::GetParam<int>("number_of_classes") == 0 ?
+        arma::max(labels) + 1 : CLI::GetParam<int>("number_of_classes");
+    model = new LinearSVM<>(0, numClasses);
+  }
+
+  // Now, do the training.
+  if (CLI::HasParam("training"))
+  {
+    model->Lambda() = lambda;
+    model->Delta() = delta;
+    numClasses = model->NumClasses();
+
+    if (optimizerType == "lbfgs")
+    {
+      model->FitIntercept() = intercept;
+
+      ens::L_BFGS lbfgsOpt;
+      lbfgsOpt.MaxIterations() = maxIterations;
+      lbfgsOpt.MinGradientNorm() = tolerance;
+
+      Log::Info << "Training model with L-BFGS optimizer." << endl;
+
+      // This will train the model.
+      model->Train(trainingSet, labels, numClasses, lbfgsOpt);
+    }
+    else if (optimizerType == "psgd")
+    {
+      const double step_size = CLI::GetParam<double>("step_size");
+      const bool shuffle = !CLI::HasParam("shuffle");
+      const size_t maxIt = epochs * trainingSet.n_cols;
+
+      ens::ConstantStep decayPolicy(step_size);
+
+      #ifndef HAS_OPENMP
+      ens::ParallelSGD<ens::ConstantStep> psgdOpt(maxIt, trainingSet.n_cols,
+        tolerance, shuffle, decayPolicy);
+      #endif
+
+      // This optimizer is only compiled if OpenMP is used.
+      #ifdef HAS_OPENMP
+      ens::ParallelSGD<ens::ConstantStep> psgdOpt(maxIt, std::ceil(
+        (float)trainingSet.n_cols / omp_get_max_threads()), tolerance, shuffle,
```

Comment:
There should be a space between the `(float)` and `trainingSet.n_cols`.  For the thread share size, perhaps you can do it like this:

```
#ifdef HAS_OPENMP
size_t threads = omp_get_max_threads();
#else
size_t threads = 1;
#endif
```

Reply:
Added This.